### PR TITLE
Add FTB App import

### LINKED ISSUE #235: FTB App Import
### Role

Playing curseforge packs

### Suggestion

Import from FTB App

### Benefit

Easier to install without having to manually install mods

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

_No response_

### PR DESCRIPTION
**Pull request by Trial97**
fixes #235

Hope I did not broke any rules

### CODE REVIEW

File: launcher/modplatform/import_ftb/PackInstallTask.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include "InstanceTask.h"
+#include "PackHelpers.h"
+
+namespace FTBImportAPP {
+
+class PackInstallTask : public InstanceTask {
+    Q_OBJECT
+
+   public:
+    explicit PackInstallTask(Modpack pack) : m_pack(pack) {}
```
[the remark this answers is not in the document]
Comment (TheKodeToad) [reply launcher/modplatform/import_ftb/PackInstallTask.h]: thanks

File: launcher/modplatform/import_ftb/PackInstallTask.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#include "minecraft/PackProfile.h"
+#include "modplatform/ResourceAPI.h"
+#include "modplatform/import_ftb/PackHelpers.h"
+#include "settings/INISettingsObject.h"
+
+namespace FTBImportAPP {
+
+void PackInstallTask::executeTask()
+{
+    setStatus(tr("Copy files"));
```
Comment (TheKodeToad) [inline launcher/modplatform/import_ftb/PackInstallTask.cpp]:
```suggestion
    setStatus(tr("Copying files..."));
```

File: launcher/modplatform/import_ftb/PackInstallTask.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        return folderCopy();
+    });
+    connect(&m_copyFutureWatcher, &QFutureWatcher<bool>::finished, this, &PackInstallTask::copySettings);
+    connect(&m_copyFutureWatcher, &QFutureWatcher<bool>::canceled, this, &PackInstallTask::emitAborted);
+    m_copyFutureWatcher.setFuture(m_copyFuture);
+}
+
+void PackInstallTask::copySettings()
+{
+    setStatus(tr("Copy settings"));
```
Comment (TheKodeToad) [inline launcher/modplatform/import_ftb/PackInstallTask.cpp]:
```suggestion
    setStatus(tr("Copying settings..."));
```

File: launcher/modplatform/import_ftb/PackInstallTask.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    instanceSettings->set("InstanceType", "OneSix");
+
+    if (m_pack.jvmArgs.isValid() && !m_pack.jvmArgs.toString().isEmpty()) {
+        instanceSettings->registerSetting("OverrideJavaArgs", true);
+        instanceSettings->set("OverrideJavaArgs", true);
+        instanceSettings->registerSetting("JvmArgs", m_pack.jvmArgs.toString());
+        instanceSettings->set("JvmArgs", m_pack.jvmArgs.toString());
+    }
+
+    MinecraftInstance instance(m_globalSettings, instanceSettings, m_stagingPath);
```
Comment (TheKodeToad) [inline launcher/modplatform/import_ftb/PackInstallTask.cpp]: If you set everything after constructing `MinecraftInstance`, then registration wouldn't be needed.

File: launcher/ui/pages/modplatform/import_ftb/ListModel.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "ListModel.h"
+#include <qnamespace.h>
```
Comment (TheKodeToad) [inline launcher/ui/pages/modplatform/import_ftb/ListModel.cpp]: i don't think this is needed

**Review by TheKodeToad — CHANGES_REQUESTED** [review]

**Comment by DioEgizio** [comment]
Maybe also we should add the Linux ftbapp path to the flatpak exceptions so that the flatpak can work with this import thing

**Comment by Trial97** [comment]
I'm not familiar with flatpak stuff can you point me to were to find that?

**Comment by DioEgizio** [comment]
https://github.com/PrismLauncher/PrismLauncher/blob/b9568279dc364a462e9837ea13bc3c4f61abdda8/flatpak/org.prismlauncher.PrismLauncher.yml#L27

You should do something like this but for that dir

See https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access

**Comment by Trial97** [comment]
Hope this works

File: launcher/modplatform/import_ftb/PackHelpers.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    try {
+        auto doc = Json::requireDocument(instanceFile.absoluteFilePath(), "FTB_APP instance JSON file");
+        const auto root = doc.object();
+        modpack.uuid = Json::requireString(root, "uuid", "uuid");
+        modpack.id = Json::requireInteger(root, "id", "id");
+        modpack.versionId = Json::requireInteger(root, "versionId", "versionId");
+        modpack.name = Json::requireString(root, "name", "name");
+        modpack.version = Json::requireString(root, "version", "version");
+        modpack.mcVersion = Json::requireString(root, "mcVersion", "mcVersion");
+        modpack.jvmArgs = Json::requireVariant(root, "jvmArgs", "jvmArgs");
```
Comment (TheKodeToad) [inline launcher/modplatform/import_ftb/PackHelpers.cpp]: This will throw if `jvmArgs` is null. i'm guessing you don't want that!

**Review by TheKodeToad — APPROVED** [review]
looks pretty good! I just noticed a potential oversight

File: launcher/modplatform/import_ftb/PackHelpers.cpp (hunk 6 above)
Comment (Trial97) [reply launcher/modplatform/import_ftb/PackHelpers.cpp]: Thanks for the catch.
Solved but right now I do not see my commits on the PR(they are on the branch) so I think maybe GitHub has a hiccup.

**Review by DioEgizio — APPROVED** [review]
Seems to work! Who is gonna update the flatpak when 8.0 will be out will have to remember to update the manifest

File: launcher/ui/pages/modplatform/import_ftb/ImportFTBPage.h [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+class ImportFTBPage;
+}
+
+class ImportFTBPage : public QWidget, public BasePage {
+    Q_OBJECT
+
+   public:
+    explicit ImportFTBPage(NewInstanceDialog* dialog, QWidget* parent = 0);
+    virtual ~ImportFTBPage();
+    QString displayName() const override { return "FTB App Import"; }
```
Comment (Scrumplex) [inline launcher/ui/pages/modplatform/import_ftb/ImportFTBPage.h]:
```suggestion
    QString displayName() const override { return tr("FTB App Import"); }
```

**Comment by Scrumplex** [comment]
> Seems to work! Who is gonna update the flatpak when 8.0 will be out will have to remember to update the manifest

flathub/org.prismlauncher.PrismLauncher#34

**Review by Scrumplex — APPROVED** [review]

**Pull request merged**